ingest/anndata_.py

```diff
@@ -35,3 +44,65 @@ def validate(self):
         except ValueError:
             return False
 
+    @staticmethod
+    def generate_cluster_header(adata, clustering_name):
+        """
+        Based on clustering dimensions, write clustering NAME line to file
+        """
+        dim2 = ['NAME', 'X', 'Y']
+        dim3 = ['NAME', 'X', 'Y', 'Z']
+        if adata.obsm[clustering_name].shape[1] == 2:
+            headers = dim2
+        elif adata.obsm[clustering_name].shape[1] == 3:
+            headers = dim3
+        elif adata.obsm[clustering_name].shape[1] > 3:
+            msg = f"Data in obsm \"{clustering_name}\" has too many dimensions for visualization."
+            # !!! is ValueError the right type?
+            raise ValueError(msg)
+        else:
+            msg = f"Data in obsm \"{clustering_name}\" has too few dimensions for visualization."
+            raise ValueError(msg)
+        with open(f"{clustering_name}.tsv", "w") as f:
+            f.write('\t'.join(headers) + '\n')
+
+    @staticmethod
+    def generate_cluster_type_declaration(adata, clustering_name):
+        """
+        Based on clustering dimensions, write clustering TYPE line to file
+        """
+        types = ["TYPE", *["numeric"] * adata.obsm['X_umap'].shape[1]]
```

Comment:
It took me longer than it should to understand this line.  Unpacking it a bit would ease maintainability, I think.  

Maybe something like:

```suggestion
        num_columns = adata.obsm["X_umap"].shape[1]
        types = ["TYPE", ["numeric"] * num_columns]
```

What does the asterisk _before_ `["numeric"]` do?  At a glance, it seems extraneous, e.g.:
```
# Without preceding asterisk 
>>> ["TYPE", ["foo"] * 3]
['TYPE', ['foo', 'foo', 'foo']]

# Same output with preceding asterisk
>>> ["TYPE", *["foo"] * 3]
['TYPE', 'foo', 'foo', 'foo']
```

P.S.: Do Black / flake8 flag such mixed quotes?  It's basically irrelevant, just a bit visually distracting for review / maintenance.

Reply:
`9bc7f61` The asterisk before `["numeric"]` unpacks the list. Without the asterisk, the ["numeric"] strings in `types` would be in an embedded list:

> ['TYPE', ['numeric', 'numeric', 'numeric']]

instead of

> ['TYPE', 'numeric', 'numeric', 'numeric']

Reply:
Oh wow, I totally overlooked that!  Thanks for pointing it out.

Reply:
I had a similar moment of confusion on this before I decided that it was in fact a splat operator, which then made sense.